### PR TITLE
Remove lodash as a production dependency

### PR DESCRIPTION
**Pull request by calebmer**
Removes `lodash` as a production dependency to reduce size. Many of the utilities were inlined such as `isString` and `isNumber`. For more involved utilities, I created a smaller version in `src/util` and added tests. These utilities include `isEqual`, `cloneDeep`, and `assign`. The smaller version in `src/util` do not cover all the edge cases that there Lodash equivalents cover. Most notably they do not handle circular objects gracefully (let me know if this is important).

Removing Lodash takes us from 136 kB to 94 kB ungzipped and 35.3 kB to 24.5 kB gzipped. Roughly a 30% decrease in size :tada:

- [x] Make sure all of the significant new logic is covered by tests
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change
- [x] Add your name and email to the AUTHORS file (optional)

### CODE REVIEW

File: src/ApolloClient.ts [hunk 1]
```diff
@@ -435,7 +432,7 @@ export default class ApolloClient {
     }
 
     // ensure existing store has apolloReducer
-    if (isUndefined(reduxRootSelector(store.getState()))) {
+    if (reduxRootSelector(store.getState()) === undefined) {
```
Comment (stubailo) [inline src/ApolloClient.ts]: The better way to do this is typeof x === 'undefined' right? Since undefined isn't a keyword and could be reassigned.

Comment (helfer) [inline src/ApolloClient.ts]: Yes, that's right in theory. I wonder if it's ever made a difference for anyone in practice?

Comment (stubailo) [inline src/ApolloClient.ts]: IDK maybe it's just one of those old JavaScript trivia things that nobody cares about anymore then.

Comment (calebmer) [inline src/ApolloClient.ts]: I think it's still a valid thing to change. The only reason I didn't do that in the first place is because lodash implements it this way: https://github.com/lodash/lodash/blob/4.17.4/lodash.js#L12212

…but they do also have this which I missed: https://github.com/lodash/lodash/blob/4.17.4/lodash.js#L12

Comment (helfer) [inline src/ApolloClient.ts]: oh, interesting (L12). I wonder why they did that instead of using typeof.

**Comment by helfer** [comment]
Great stuff @calebmer! 

I'd say that's enough low-hanging fruit so we can close the other issue. Except maybe that we want to get rid of whatwg-fetch as well for other reasons.

img

**Comment by stubailo** [comment]
Gosh guys this is incredible. Great work @calebmer!

**Comment by stubailo** [comment]
BTW other low-hanging fruit: getFromAST and storeUtils are duplicated in apollo-client and graphql-anywhere both.
